### PR TITLE
ESQL: Finish a TODO in BlockHashes

### PR DESCRIPTION
**Pull request by nik9000**
We had a TODO in our `BlockHash` implementations optimized for pairs of columns - we wanted to use `MultiValueDedupe` inside their `add` methods for `Block`s. This implements to TODO.

It makes a small behavior change on one of the blocks we don't yet use in production - the `BytesRefLongBlockHash` will now properly keep `null` bytes strings. That's a side effect of reusing other components and doesn't actually allow us to use it in production - we're still waiting on the joint hash tables which are blocked behind vector instructions.

I moved a bunch of files to different places so I could reach into the innards of the `MultivalueDedupe` subclasses to build the block hash addition logic. It seemed like a reasonable thing to do. And it seemed reasonable not to expose the raw arrays outside of the package.

### CODE REVIEW

File: benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/AggregatorBenchmark.java [hunk 1]
```diff
@@ -575,13 +575,14 @@ private static void run(String grouping, String op, String blockType, int opCoun
...
 
         DriverContext driverContext = driverContext();
-        Operator operator = operator(driverContext, grouping, op, dataType);
-        Page page = page(driverContext.blockFactory(), grouping, blockType);
-        for (int i = 0; i < opCount; i++) {
-            operator.addInput(page);
+        try (Operator operator = operator(driverContext, grouping, op, dataType)) {
+            Page page = page(driverContext.blockFactory(), grouping, blockType);
+            for (int i = 0; i < opCount; i++) {
+                operator.addInput(page.shallowCopy());
```
Comment (dnhatn) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/AggregatorBenchmark.java:581]: do we need to call `shallowCopy`?

**Review by dnhatn — APPROVED** [review]
LGTM, thanks Nik!

File: benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/AggregatorBenchmark.java (hunk 1 above)
Comment (nik9000) [reply benchmarks/src/main/java/org/elasticsearch/benchmark/compute/operator/AggregatorBenchmark.java:581]: Yeah, we build the page one time outside the loop. Without that it crashes on the second run.

**Comment by nik9000** [comment]
> @nik9000 Sorry about the merge conflicts. I didn't realize we were working on the same files; otherwise, I would've postponed merging my PR.

No problem!

**Pull request merged**